### ISSUE TITLE
contents local does not include content of 'only' directive

### ISSUE THREAD

**Issue by ArtFlag** [sphinx-doc/sphinx#4403]
### Problem

```
.. contents::
   :local:
```
Does not generate links to titles that are inside an "only" block.

#### Procedure to reproduce the problem

Content:

```
Title
=====

.. contents::
   :local:

Heading1
--------

Bla

.. only:: internal

   Heading2
   --------

   Blabala

```

Output:
![screen shot 2018-01-09 at 11 34 00](https://user-images.githubusercontent.com/25686502/34716934-9fe2e4ba-f531-11e7-8b98-d3fbbad1c7bd.png)

#### Expected results
![screen shot 2018-01-09 at 11 34 59](https://user-images.githubusercontent.com/25686502/34716936-a3b2e464-f531-11e7-9f48-659800a68316.png)

### Environment info
- OS: High Sierra
- Sphinx version: 1.6.5

**Comment by ArtFlag** [comment]
Hi, 

Can we tag this as a bug, unless I'm wrong?

**Comment by jfbu** [comment]
Thanks for reporting, I agree this is disappointing but Sphinx documentation contains [this](http://www.sphinx-doc.org/en/master/markup/misc.html#directive-only)

> Warning
> 
> This directive is designed to control only content of document. It could not control sections, labels and so on.
> 

Hence it does not look as if your finding qualifies as bug. At any rate though, documentation should be even clearer.

**Comment by tk0miya** [comment]
Certainly, I also consider this is not a bug. But it is not unexpected behavior.

Note:
This comes evaluation timing of contents directive and only directive. First one is evaluated on reading phase, and the another is on resolving phase.
This would be fixed if we extract content node (accurately topic node) on writing node.
